SonarQube.Client/Helpers/SecondaryIssueHashUpdater.cs

```diff
@@ -0,0 +1,107 @@
+﻿/*
+ * SonarQube Client
+ * Copyright (C) 2016-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using SonarQube.Client.Models;
+
+namespace SonarQube.Client.Helpers
+{
+    /// <summary>
+    /// Sets the hashes for any secondary locations in the supplied list of issues
+    /// </summary>
+    /// <remarks>
+    /// Currently secondary location hashes are not stored server-side, so we have to
+    /// calculate them ourselves. This means fetching the source code for each file
+    /// so we can get the line text and calculate the hash.
+    /// </remarks>
+    internal class SecondaryIssueHashUpdater
+    {
+        private static readonly IChecksumCalculator checksumCalculator = new ChecksumCalculator();
+        private Dictionary<string, string> moduleKeyToSourceMap;
+
+        internal async Task UpdateHashesAsync(IEnumerable<SonarQubeIssue> issues,
+            ISonarQubeService sonarQubeService,
+            CancellationToken cancellationToken)
+        {
+            var uniqueKeys = GetUniqueSecondaryLocationKeys(issues);
```

Comment:
Minor optimization: we can call `GetSecondaryLocations` only once and pass `IEnumerable<IssueLocation>` to GetUniqueSecondaryLocationKeys.